### DISCUSSION TITLE
Echo while using the duplex pass in LiveEffect App

### DISCUSSION THREAD

**Discussion by gitvignesh** [google/oboe#1958] [Q&A] [answered]
Hi,
when we run the live effect app by attaching an external headphone with a mic to a device and try to increase the volume of the same the audio I hear back seems to echo/distort beyond a volume point. (The audio is not looping back physically)
Can someone explain to me why this is so?

**Comment by robertwu1** [comment]
There is probably no Acoustic Echo Canceler (AEC) in your case so there is a feedback loop.

**Reply by gitvignesh** [reply]
But if I am looping back the audio to the earphones and there is no leakage of audio back to the mic, why do I need an AEC?

**Reply by robertwu1** [reply]
It seems like instead of this being a feedback loop, your output volume is so high that it starts clipping.

See [https://github.com/google/oboe/blob/main/samples/LiveEffect/src/main/cpp/FullDuplexPass.h](https://github.com/google/oboe/blob/main/samples/LiveEffect/src/main/cpp/FullDuplexPass.h).
~~~
        for (int32_t i = 0; i < samplesToProcess; i++) {
            *outputFloats++ = *inputFloats++ * 0.95; // do some arbitrary processing
        }
~~~

Which phone are you testing this on? I don't see this issue with my devices. Perhaps your phone has some special feature where the volume goes above 1.

**Reply by gitvignesh** [reply]
I have a physical volume rocker on my headphones, until the rocker is below 60% everything sounds okay but as soon as the rocker passes 60% everything gets distorted. Also, the mic picks up the far noise more than the closer noises.

**Comment by philburk** [comment]
What were the devices? USB headset model? Phone?

There must be some feedback somewhere in the system. There is some electrical coupling between the wires of the headset cable but that should be very small.

What happens if you use the "ECHO INPUT TO OUTPUT" test in OboeTester?

There is also the RECORD AND PLAY TEST that does each separately so there should be no feedback possible.

**Reply by gitvignesh** [reply]
Android device: Samsung Galaxy M31
Headset: JBL Quantum 100 Wired
Test Output: 
cold.start.input.msec = 157
cold.start.output.msec = 191
stable.call.count = 21

I have a physical volume rocker on my headphones, until the rocker is below 60% everything sounds okay but as soon as the rocker passes 60% everything gets distorted. Also, the mic picks up the far noise more than the closer noises.

**Comment by philburk** [comment]
When you increase the volume above a certain level then does the system go into a sustained feedback mode even when you are not talking? This would be an echo problem. The solution would involve finding the source of the feedback.

Or when you increase the volume does the sound just sound clipped and distorted when someone is talking and then become quiet when no one is talking? That would be a problem with non-linear distortion in the output path. I have seen this before. The distortion may be coming from the processing in the DSP, or the analog circuitry, or even in the speakers in the headphone. In this case there may be no solution except keeping the volume low or trying different headphones?

>  Also, the mic picks up the far noise more than the closer noises.

The M31 may be running an AGC to boost quiet sounds. It may need tuning.

**Reply by gitvignesh** [reply]
whenever I increase the volume above a certain level even if I don't talk from the beginning there is a lot of sustained distorted noise, but as soon as I drop back the gain it becomes normal.

> The M31 may be running an AGC to boost quiet sounds. It may need tuning.

Is there a way to disable or tune this while recording with Oboe?

**Comment by philburk** [comment]
> whenever I increase the volume above a certain level even if I don't talk from the beginning there is a lot of
> sustained distorted noise, but as soon as I drop back the gain it becomes normal.

OK, that sounds like a positive feedback loop. If the loop gain goes over 1.0 then you will get a sustained distorted noise.
Normally there is not enough signal going from the headset speakers to the headset mic to cause sustained feedback.
So the mic gain must be very high.

Is there any sound coming from the speakers on the phone? Or is all coming from the headset?

Also try covering up the microphone on the headset with your finger. The mic is probably in a tiny hole near the volume control.
Does that prevent the feedback?


> Is there a way to disable or tune this while recording with Oboe?

Maybe. Try this experiment with "ECHO INPUT TO OUTPUT" in OboeTester.
Press START and turn up the volume. Does it feedback?

Then STOP and press the green bars to exposes the setings.
Set the Output Usage to "VoiceComm"
and the Input Preset to "VoiceComm".

Then press START and turn up the volume. Does it feedback?

**Reply by gitvignesh** [reply]
>Is there any sound coming from the speakers on the phone? Or is all coming from the headset?
Also try covering up the microphone on the headset with your finger. The mic is probably in a tiny hole near the volume control.
Does that prevent the feedback?

No, I have a boom mic attached to the headset, even though I cover the Mic there is distortion when volume goes beyond the limit.

>Maybe. Try this experiment with "ECHO INPUT TO OUTPUT" in OboeTester.
Press START and turn up the volume. Does it feedback?
Then STOP and press the green bars to exposes the setings.
Set the Output Usage to "VoiceComm"
and the Input Preset to "VoiceComm".

Yes, But if we set VoiceComm there is no feedback.

**Comment by philburk** [comment, marked as answer]
I suspect the feedback is caused by an overly aggressive automatic gain control in the headset or in the phone.
An AGC can amplify even the tiniest sound to an audible level.

Using VoiceComm will eliminate the feedback because it adds an Automatic Echo Cancellation (AEC) effect.
